### PR TITLE
[Workers] Chat tutorial with durable objects

### PR DESCRIPTION
**Pull request by deadlypants1973**
Not ready for review until Cody K. updates the repository for Wrangler 2

### CODE REVIEW

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+---
+
+# Deploy a real-time chat application
+
+In this tutorial, you will deploy a serverless, real-time chat application that runs using [Durable Objects](/workers/learning/using-durable-objects). 
+
+This chat application uses a Durable Object to control each chat room. Users connect to the Object using WebSockets. Messages from one user are broadcast to all the other users. The chat history is also stored in durable storage. Real-time messages are relayed directly from one user to others without going through the storage layer.
+
+To continue with this tutorial, you must:
+1. Install [Wrangler](/workers/wrangler/get-started).
```
Comment (koeninger) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]: Probably worth specifying Wrangler 2

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+```sh
+$ wrangler login
+```
+
+## Publish your project
+
+Open the cloned workers-chat-demo repository. To test the chat application in a local environemnt, run:
+
+```sh
+$ wrangler dev
```
Comment (koeninger) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]: wrangler dev doesn't work currently, not clear to me whether it ever did, console shows

```
(index):371 WebSocket connection to 'wss://127.0.0.1:8787/api/room/some-room/websocket' failed: 
```

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+After you have cloned the repository, authenticate Wrangler by running:
+
+```sh
+$ wrangler login
+```
+
+## Publish your project
+
+Open the cloned workers-chat-demo repository. To test the chat application in a local environemnt, run:
```
Comment (koeninger) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]: "local environemnt" -> "local environment"

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+To configure a subdomain in your `wrangler.toml`, add the following to your environment:
+
+```toml
+Route = [
+{pattern = “subdomain.example.com”, 
+custom domains = true,
+}]
+```
+
+By completing this tutorial, you have deployed a real-time chap application with Durable Objects and Cloudflare Workers.
```
Comment (koeninger) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]: "real-time chap" -> "real-time chat"

**Comment by koeninger** [comment]
may seem super obvious, but I think it's worth describing how to actually use the thing - e.g. go to edge-chat-demo.your-name.workers.dev in your browser, put in your name and a chat room, send some messages, try the same room url from a different browser / different name

Also think it's worth linking the deletion instructions on the readme

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md (hunk 1 above)
Comment (deadlypants1973) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]:
```suggestion
1. Install [Wrangler 2](/workers/wrangler/get-started).
```

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md (hunk 3 above)
Comment (deadlypants1973) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]:
```suggestion
Open the cloned workers-chat-demo repository. To test the chat application in a local environment, run:
```

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+```
+
+## Publish your project
+
+Open the cloned workers-chat-demo repository. To test the chat application in a local environemnt, run:
+
+```sh
+$ wrangler dev
+```
+
```
Comment (deadlypants1973) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]:
```suggestion
```

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md (hunk 4 above)
Comment (deadlypants1973) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]:
```suggestion
By completing this tutorial, you have deployed a real-time chat application with Durable Objects and Cloudflare Workers.
```

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+updated: 2022-07-21
+difficulty: Intermediate
+content_type: 📝 Tutorial
+pcx-content-type: tutorial
+title: Deploy a real-time chat application
+---
+
+# Deploy a real-time chat application
+
+In this tutorial, you will deploy a serverless, real-time chat application that runs using [Durable Objects](/workers/learning/using-durable-objects).
```
Comment (marciocloudflare) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]:
```suggestion
In this tutorial, you will deploy a serverless, real-time chat application that runs using [Durable Objects](/workers/learning/using-durable-objects/). 
```

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+---
+
+# Deploy a real-time chat application
+
+In this tutorial, you will deploy a serverless, real-time chat application that runs using [Durable Objects](/workers/learning/using-durable-objects). 
+
+This chat application uses a Durable Object to control each chat room. Users connect to the Object using WebSockets. Messages from one user are broadcast to all the other users. The chat history is also stored in durable storage. Real-time messages are relayed directly from one user to others without going through the storage layer.
+
+To continue with this tutorial, you must:
+1. Install [Wrangler 2](/workers/wrangler/get-started).
```
Comment (marciocloudflare) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]:
```suggestion
1. Install [Wrangler 2](/workers/wrangler/get-started/).
```

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+```sh
+$ wrangler login
+```
+
+## Publish your project
+
+Open the cloned workers-chat-demo repository. To test the chat application in a local environment, run:
+
+When you are ready to deploy your application, run:
```
Comment (marciocloudflare) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]: Not sure if there's a line missing after line 37 or if we should delete line 39

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+new_classes = ["ChatRoom", "RateLimiter"]
+
+[[migrations]]
+tag = "v2"
+deleted_classes = ["ChatRoom", "RateLimiter"]
+```
+
+Then run `wrangler publish`.
+
+To delete your Worker, log in to the [Cloudflare dashboard](https://dash.cloudflare.com) > **Workers** > your Worker > **Manage Service** on the right side of the screen > **Delete**. For complete instructions on set up and deletion, refer to the `README.md` in your cloned repository.
```
Comment (marciocloudflare) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]: You probably should avoid having `on the right side of the screen` in this sentence, as the interface might change.

**Review by marciocloudflare — APPROVED** [review]

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+After you have cloned the repository, authenticate Wrangler by running:
+
+```sh
+$ wrangler login
+```
+
+## Publish your project
+
+Open the cloned workers-chat-demo repository. To test the chat application in a local environment, run:
```
Comment (deadlypants1973) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]:
```suggestion
```

File: content/workers/tutorials/deploy-a-realtime-chat-app/index.md (hunk 9 above)
Comment (deadlypants1973) [inline content/workers/tutorials/deploy-a-realtime-chat-app/index.md]:
```suggestion
To delete your Worker, log in to the [Cloudflare dashboard](https://dash.cloudflare.com) > **Workers** > your Worker > **Manage Service** > **Delete**. For complete instructions on set up and deletion, refer to the `README.md` in your cloned repository.
```

**Review by koeninger — APPROVED** [review]

**Pull request merged**